lib/auth/bot.go

```diff
@@ -281,6 +361,52 @@ func (a *Server) updateBotInstance(
 		return nil
 	}
 
+	l := log.WithFields(logrus.Fields{
+		"bot_name":        botName,
+		"bot_instance_id": botInstanceID,
+	})
+
+	if currentIdentityGeneration == 0 {
+		// Nothing to do.
+		l.Warn("bot attempted to fetch certificates without providing a current identity generation, this is not allowed")
+
+		return trace.AccessDenied("a current identity generation must be provided")
+	} else if currentIdentityGeneration > 0 && currentIdentityGeneration != instanceGeneration {
+		// For now, continue to only enforce generation counter checks on
+		// renewable (i.e. token) identities.
+		if req.renewable {
+			if err := a.tryLockBotDueToGenerationMismatch(ctx, username); err != nil {
+				l.WithError(err).Warn("Failed to lock bot when a generation mismatch was detected")
+			}
+
+			return trace.AccessDenied(
+				"renewable cert generation mismatch for bot %s/%s: stored=%v, presented=%v",
+				botName, botInstanceID,
+				instanceGeneration, currentIdentityGeneration,
+			)
+		} else {
+			// We'll still log the check failure, but won't deny access. This
+			// log data will help make an informed decision about reliability of
+			// the generation counter for all join methods in the future.
+			l.WithFields(logrus.Fields{
+				"bot_instance_generation": instanceGeneration,
+				"bot_identity_generation": currentIdentityGeneration,
+				"bot_join_method":         authRecord.JoinMethod,
+			}).Warn("Bot generation counter mismatch detected. This check is not enforced for this join method, " +
+				"but may indicate multiple uses of a bot identity and possibly a compromised certificate.")
```

Comment:
I am curious for feedback on whether or not this should be warning level. We definitely want it for our own analysis (Cloud logs, etc) but it might be spammy for self hosted users. It _could_ indicate stolen certs (unlikely), or just a data dir mishap.

Reply:
It is genuinely indicative of a failure, and one that we want to learn more about / eradicate. Whilst it's not immediately actionable for users, if they get in touch, we may be able to learn something about why their cluster is specifically encountering this more often and do something in response to that. So I don't think warning is necessarily too bad.

Reply:
Fair enough, I'm happy to keep it as is if there's no objections to the verbiage.